### PR TITLE
Generated types should be annotated with @Generated #178

### LINKED ISSUE #178: Generated types should be annotated with javax.annotation.processing.Generated
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
Any generated type (class, interface, enum, ...) should be annotated with [`Generated`](https://docs.oracle.com/en/java/javase/11/docs/api/java.compiler/javax/annotation/processing/Generated.html).

**Describe alternatives you've considered**
I have not considered any alternative solutions.

**Additional context**
Other code generation libraries do this. Examples:

##### MapStruct

```java
@Generated(
  value = "org.mapstruct.ap.MappingProcessor",
  date = "2020-06-16T13:43:26+0200",
  comments = "version: 1.3.1.Final, compiler: javac, environment: Java 11.0.2 (Oracle Corporation)"
)
public class MyMapperImpl implements MyMapper {}
```

##### Hibernate JPA Model Generation

```java
@Generated(value = "org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor")
@StaticMetamodel(MyEntity.class)
public abstract class MyEntity_  {}
```

##### OpenAPI Tools

See [here](https://github.com/OpenAPITools/openapi-generator/blob/32adeddd8fd6e217eca2092f4bf370e6d6e7ea33/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/Pig.java#L58).

### PR DESCRIPTION
**Pull request by kobylynskyi**
Resolves #178

### CODE REVIEW

**Comment by kobylynskyi** [comment]
@lorenzleutgeb @joffrey-bion could you guys please take a look?

File: src/main/java/com/kobylynskyi/graphql/codegen/model/DefaultDateTimeGenerator.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.kobylynskyi.graphql.codegen.model;
+
+public class DefaultDateTimeGenerator implements DateTimeGenerator {
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/DefaultDateTimeGenerator.java:3]:
```suggestion
public class DefaultDateTimeGenerator implements Supplier<ZonedDateTime> {
    public ZonedDateTime get() {
        return ZonedDateTime.now();
    }
```

File: src/main/java/com/kobylynskyi/graphql/codegen/model/DefaultDateTimeGenerator.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.kobylynskyi.graphql.codegen.model;
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/DefaultDateTimeGenerator.java:1]:
```suggestion
package com.kobylynskyi.graphql.codegen.model;

import java.time.ZonedDateTime;
import java.util.function.Supplier;
```

File: src/main/java/com/kobylynskyi/graphql/codegen/model/DateTimeGenerator.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.kobylynskyi.graphql.codegen.model;
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/DateTimeGenerator.java:1]: Delete this?

File: src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package com.kobylynskyi.graphql.codegen.model;
+
+import java.time.format.DateTimeFormatter;
+
+public class GeneratedInformation {
+
+    public static final DateTimeFormatter DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssZ");
+
+    private DateTimeGenerator dateTimeGenerator;
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java:9]:
```suggestion
    private Supplier<ZonedDateTime> dateTimeSupplier;
```

File: src/test/java/com/kobylynskyi/graphql/codegen/TestUtils.java [hunk 5]
```diff
@@ -21,7 +27,33 @@ public static File getFileByName(File[] files, String fileName) throws FileNotFo
...
+    }
+
+    public static GeneratedInformation getStaticGeneratedInfo() {
+        GeneratedInformation generatedInformation = new GeneratedInformation();
+        generatedInformation.setDateTimeGenerator(new DateTimeGenerator() {
+            @Override
+            public ZonedDateTime newDateTime() {
+                return GENERATED_DATE_TIME;
+            }
+        });
```
Comment (lorenzleutgeb) [inline src/test/java/com/kobylynskyi/graphql/codegen/TestUtils.java:55]:
```suggestion
        generatedInformation.setDateTimeGenerator(() -> GENERATED_DATE_TIME);
```

File: src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    private final String generatedType;
+
+    public GeneratedInformation() {
+        this(new DefaultDateTimeGenerator());
+    }
+
+    public GeneratedInformation(DateTimeGenerator dateTimeGenerator) {
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java:16]:
```suggestion
    public GeneratedInformation(Supplier<ZonedDateTime> dateTimeGenerator) {
```

File: src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        this.dateTimeGenerator = dateTimeGenerator;
+        this.generatedType = initGeneratedType();
+    }
+
+    public void setDateTimeGenerator(DateTimeGenerator dateTimeGenerator) {
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java:21]:
```suggestion
    public void setDateTimeGenerator(Supplier<ZonedDateTime> dateTimeGenerator) {
```

File: src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        this.dateTimeGenerator = dateTimeGenerator;
+    }
+
+    public String getGeneratedType() {
+        return generatedType;
+    }
+
+    public String getDateTime() {
+        return DATE_TIME_FORMAT.format(dateTimeGenerator.newDateTime());
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java:30]:
```suggestion
        return DATE_TIME_FORMAT.format(dateTimeGenerator.get());
```

**Comment by lorenzleutgeb** [comment]
Side question: What's the source code Java version of this project?

File: src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+public class GeneratedInformation {
+
+    public static final DateTimeFormatter DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssZ");
+
+    private DateTimeGenerator dateTimeGenerator;
+    private final String generatedType;
+
+    public GeneratedInformation() {
+        this(new DefaultDateTimeGenerator());
```
Comment (lorenzleutgeb) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/GeneratedInformation.java:13]:
```suggestion
        this(ZonedDateTime::now);
```

If this is the only use of `DefaultDateTimeGenerator`, then maybe remove it altogeher.

**Comment by kobylynskyi** [comment]
@lorenzleutgeb, thanks for your valuable input.
I've changed the implementation based on your suggestion. 
This removed both `DateTimeGenerator` and `DefaultDateTimeGenerator` at all.
And regarding the source code version - right now it is 8. But I am making the build working on the higher versions too.

**Pull request merged**